plugins/pipelines/app/models/concerns/samson_pipelines/stage_concern.rb

```diff
@@ -0,0 +1,37 @@
+module SamsonPipelines::StageConcern
+  # Return true if any stages in the pipeline are marked production
+  def production?
+    super || next_stages.any?(&:production?)
+  end
+
+  def next_stage
+    next_stage_ids.empty? ? super : Stage.find(next_stage_ids.first)
+  end
+
+  def next_stages
+    Stage.find(next_stage_ids)
+  end
+
+  protected
+
+  # Ensure we don't have a circular pipeline:
+  #
+  # potential race-condition if 2 stages are saved at same time:
+  #   stageA saved with pipelines to stageB and stageC
+  #   stageC saved with pipeline to stageA   => will validate if stageA above hasn't been written to DB yet
+  def valid_pipeline?(origin_id = id)
+    Rails.logger.warn "valid_pipeline? #{origin_id} / #{id} - #{next_stage_ids}"
+    if next_stage_ids.any? { |next_id| next_id.to_i == origin_id.to_i }
+      errors[:base] << "Stage #{name} causes a circular pipeline with this stage"
+      return false
+    end
+
+    next_stages.each do |stage|
+      unless stage.valid_pipeline?(origin_id)
+        errors[:base] << "Stage #{stage.name} causes a circular pipeline with this stage"
+        return false
```

Comment:
I need it to return false as it's called recursively down the chain of linked stages.